### PR TITLE
Create output parent directory when necessary

### PR DESCRIPTION
**Pull request by arnaudcourtecuisse**
Implement #408 with tests

### CODE REVIEW

**Comment by domoritz** [comment]
I think this should be the default. Let's not add an option.

**Review by domoritz — CHANGES_REQUESTED** [review]
I think this should be the default. Let's not add an option.

File: typescript-json-schema.ts [hunk 1]
```diff
@@ -1677,12 +1688,28 @@ export function exec(filePattern: string, fullTypeName: string, args = getDefaul
 
     const json = stringify(definition, { space: 4 }) + "\n\n";
     if (args.out) {
-        require("fs").writeFile(args.out, json, function (err: Error) {
-            if (err) {
-                throw new Error("Unable to write output file: " + err.message);
+        const fs = require("fs");
+        fs.writeFile(args.out, json, function (err: Error) {
```
Comment (domoritz) [inline typescript-json-schema.ts:1692]: Let's simplify this code and always run `fs.mkdir`. It should be no-op when the dir already exists.

File: typescript-json-schema.ts [hunk 2]
```diff
@@ -1653,7 +1653,18 @@ function normalizeFileName(fn: string): string {
...
+    if (err !== null) {
+        throw err;
+    }
+}
+
+export function exec(
+    filePattern: string,
+    fullTypeName: string,
+    args = getDefaultArgs(),
+    callback: (err: Error | null) => void = defaultCallback
```
Comment (domoritz) [inline typescript-json-schema.ts:1666]: Let's not add a callback. Just throw an error like before.

**Review by domoritz — CHANGES_REQUESTED** [review]

File: typescript-json-schema.ts (hunk 2 above)
Comment (domoritz) [reply typescript-json-schema.ts:1666]: If you want, you can mark the function as async in a follow up pull request. I prefer promises over callbacks.

Comment (arnaudcourtecuisse) [reply typescript-json-schema.ts:1666]: The callback is necessary to trigger the tests in `test/out-option.test.ts`. The only alternative I see would be to use `fs.writeFileSync` and `fs.mkdirSync`. Do you want me to do so ? I'm open to suggestions if you have other ideas.

Comment (domoritz) [reply typescript-json-schema.ts:1666]: I see. Yes, let's use the sync versions for now.

Comment (arnaudcourtecuisse) [reply typescript-json-schema.ts:1666]: I prefer promises to. I useda callback to minimize the change.

Let's put this PR on hold. I'll open another to use a promise, and once the promise is merged, I'll update this one.

Comment (domoritz) [reply typescript-json-schema.ts:1666]: Sounds good. Thanks!

**Comment by arnaudcourtecuisse** [comment]
I rebased my branch and used a promise instead of a callback

**Pull request merged**